spark/src/main/scala/org/apache/spark/sql/delta/commands/merge/ClassicMergeExecutor.scala

```diff
@@ -101,9 +99,9 @@ trait ClassicMergeExecutor extends MergeIntoMaterializeSource with MergeOutputGe
     val matchedPredicate =
       if (isMatchedOnly) {
         matchedClauses
-          .map(clause => clause.condition.getOrElse(Literal(true)))
-          .reduce((a, b) => Or(a, b))
-      } else Literal(true)
+          .flatMap(_.condition)
+          .foldLeft[Expression](Literal.TrueLiteral)(Or.apply)
```

Comment:
"flatMap will ignore any clause that doesn't have a condition whereas the result should be true if one of the clauses doesn't have a condition." and that's exactly why `foldLeft` starts with...wait...I think I found a bug in my "simplification"! 😄

Reply:
There's still an issue with:
```
    val matchedPredicate =
      if (isMatchedOnly) {
        matchedClauses
          .flatMap(_.condition)
          .foldLeft[Expression](Literal.FalseLiteral)(Or.apply)
      } else Literal.TrueLiteral
```

If we have a single matched clause without condition then this will return `Literal.FalseLiteral` when we should get `Literal.TrueLiteral` since all rows qualify.

I think it's best to keep the previous expression here, the optimizer will be able to eliminate trivially true predicates